src/core/lib/event_engine/workqueue.h

```diff
@@ -0,0 +1,238 @@
+// Copyright 2022 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GRPC_CORE_LIB_EVENT_ENGINE_WORKQUEUE_H
+#define GRPC_CORE_LIB_EVENT_ENGINE_WORKQUEUE_H
+
+#include <grpc/support/port_platform.h>
+
+#include <stddef.h>
+
+#include <atomic>
+#include <deque>
+#include <utility>
+
+#include "absl/base/internal/spinlock.h"
+#include "absl/base/thread_annotations.h"
+#include "absl/cleanup/cleanup.h"
+#include "absl/synchronization/mutex.h"
+#include "absl/types/optional.h"
+
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/gprpp/time.h"
+#include "src/core/lib/iomgr/exec_ctx.h"
+
+namespace grpc_event_engine {
+namespace experimental {
+
+// A fast work queue based lightly on an internal Google implementation.
+//
+// This uses atomics and lightweight spinlocks to access the most recent element
+// in the queue, making it fast for LIFO operations. Accessing the oldest (next)
+// element requires taking a mutex lock.
+template <typename T>
+class WorkQueue {
```

Comment:
Any chance we could make this non-templated?

Reply:
We can. I made it templated in case we wanted to support AnyInvocable and Closure* types. We could specialize the queue to one or the other, or both with overloads/separate methods. Today's thread pool only uses AnyInvocables, so we could start there.

Is there any particular reason you'd want to avoid templates in this code, specifically?

Reply:
I specialized the queue for Closure & AnyInvocable. There's some asymmetry in dealing with the two types given that Closure* callbacks are not owned by the queue, but AnyInvocables are, so the common usage is that closures _must_ be executed or a leak will happen (or if the caller knows it's an Closure-wrapped AnyInvocable, then they can delete it instead).